contrib/lib/src/databases.rs

```diff
@@ -843,6 +846,82 @@ impl Poolable for memcache::Client {
     }
 }
 
+#[doc(hidden)]
+/// Unstable internal details of generated code for the #[database] attribute.
+///
+/// This type is implemented here instead of in generated code to ensure all
+/// types are properly checked.
+pub struct ConnectionPool<C: Poolable> {
+    pool: r2d2::Pool<C::Manager>,
+    semaphore: Arc<Semaphore>,
+}
+
+impl<C: Poolable> ConnectionPool<C> {
+    pub fn fairing(fairing_name: &'static str, config_name: &'static str) -> impl rocket::fairing::Fairing {
+        rocket::fairing::AdHoc::on_attach(fairing_name, move |mut rocket| async move {
+            let config = database_config(config_name, rocket.config().await);
+            let pool = config.map(|c| (c.pool_size, C::pool(c)));
+
+            match pool {
+                Ok((size, Ok(p))) => {
+                    let managed = ConnectionPool::<C> {
+                        pool: p,
+                        semaphore: Arc::new(Semaphore::new(size as usize)),
+                    };
+                    Ok(rocket.manage(managed))
+                },
+                Err(config_error) => {
+                    ::rocket::logger::error(
+                        &format!("Database configuration failure: '{}'", config_name));
+                    ::rocket::logger::error_(&format!("{}", config_error));
+                    Err(rocket)
+                },
+                Ok((_, Err(pool_error))) => {
+                    ::rocket::logger::error(
+                        &format!("Failed to initialize pool for '{}'", config_name));
+                    ::rocket::logger::error_(&format!("{:?}", pool_error));
+                    Err(rocket)
+                },
+            }
+        })
+    }
+
+    pub fn get_one(cargo: &::rocket::Cargo) -> Option<Self> {
+        cargo.state::<Self>().map(|c| Self {
+            pool: c.pool.clone(),
+            semaphore: c.semaphore.clone(),
+        })
+    }
+
+    pub async fn run<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce(&mut r2d2::PooledConnection<C::Manager>) -> R + Send + 'static,
+        R: Send + 'static,
+    {
+        let _permit = self.semaphore.acquire().await;
+        let pool = self.pool.clone();
+        let ret = tokio::task::spawn_blocking(move || {
+            let mut conn = pool.get().expect("TODO");
```

Comment:
This is... not great. By acquiring a connection in `run` instead of in `from_request`, `run()` will need to return a `Result<_, r2d2::Error>`. Database operations are also likely to return a `Result`, though, so `Result<Result<T, db::Error>, r2d2::Error>>` would be typical. Doing this would add a lot of error-handling boilerplate at usage sites that wasn't there before.